ic-agent/src/identity/mod.rs

```diff
@@ -24,19 +26,75 @@ pub struct Signature {
     pub signature: Option<Vec<u8>>,
 }
 
-/// An Identity takes a request id and returns the [Signature]. It knows or
-/// represents the Principal of the sender.
+/// An `Identity` produces [`Signatures`](Signature) for requests or delegations. It knows or
+/// represents the [`Principal`] of the sender.
 ///
-/// Agents are assigned a single Identity object, but there can be multiple
+/// [`Agents`](crate::Agent) are assigned a single `Identity` object, but there can be multiple
 /// identities used.
 pub trait Identity: Send + Sync {
     /// Returns a sender, ie. the Principal ID that is used to sign a request.
     ///
     /// Only one sender can be used per request.
     fn sender(&self) -> Result<Principal, String>;
 
+    /// Produce the public key commonly returned in [`Signature`].
+    ///
+    /// May return `None` when [`sign`](Identity::sign) would return `Some`, in non-`ic-agent` identities.
+    fn public_key(&self) -> Option<Vec<u8>>;
```

Comment:
Do we want to provide a default implementation to return `None`?

Reply:
No. The reason that is there is for signatures that differ in their public keys depending on the content being signed, *not* because it should be considered an optional function.

Reply:
Let's document that. It's unlike when this function returns None.

Reply:
Actually I think a better option is to remove it since the use-case I was envisioning is better designed differently.